### PR TITLE
Add SassyStringModifier

### PR DESCRIPTION
**Pull request by marczalik**
**One sentence summary of this PR (This should go in the CHANGELOG!)**
Sassify your strings with ChatGPT.

**Link to Related Issue(s)**
N/A

**Please describe the changes in your request.**
Add Modifier component that targets ASCII strings of a certain length and then attempts to sassify them using the power of AI.

**Anyone you think should look at this, specifically?**
@whyitfor @ANogin 

Note: This is just a draft PR at the moment as I'm still actively working on making improvements. I wanted to get this PR out there so that people could have a chance to look at the code, test it out for themselves, and suggest feedback.

### CODE REVIEW

**Comment by ANogin** [comment]
@marczalik do I read this correctly that you are asking ChatGPT to follow rules (length, spaces in identifier, string format), but you are not checking it listened? I am sure there will be cases where the rule would be violated, so you ought to check.

Also, is there some standard way to more explicitly emphasize to ChatGPT where the prompt end and the data begins, so that there is less chance ChatGPT would react to instruction-like text in the actual code string?

**Comment by marczalik** [comment]
> @marczalik do I read this correctly that you are asking ChatGPT to follow rules (length, spaces in identifier, string format), but you are not checking it listened? I am sure there will be cases where the rule would be violated, so you ought to check.
> 
> Also, is there some standard way to more explicitly emphasize to ChatGPT where the prompt end and the data begins, so that there is less chance ChatGPT would react to instruction-like text in the actual code string?

@ANogin there will definitely be cases where rules given in the prompts are violated - in fact, it happens quite frequently. That is why I compare the result length to the input length and request shorter responses up to `max_retries` times before truncating the response to force the result to fit the input length. I also pull out any non-ASCII characters manually. Unfortunately I haven't found a prompt that consistently prevents ChatGPT from adding extraneous commentary that breaks the no spaces request, hence why I make the (potentially erroneous) assumption that the longest word in the response is the sassified identifier. This may not be the best solution though. Did you have another method by which to check that the rules are followed?

I am not aware of there being a more explicit method for delineating prompt from data. So far I haven't come across any issues where the prompt is mistaken for the data, although that could happen. One issue I did see while testing is that, if I tried to give a structured example such as 
```
Request: <example input>
Response: <example output>
```
I would start to see ChatGPT responding with the structure embedded in its response. E.g., it would sometimes say
```
Response: <blah blah blah>
```
which made it even harder to predict/parse its output, hence why I removed it.

File: ofrak_ai/ofrak_ai/chatgpt.py
Comment (rbs-jacob) [inline ofrak_ai/ofrak_ai/chatgpt.py]: Have you tried speeding things up by changing this parameter? Does changing this affect the cost of the query?

File: ofrak_ai/ofrak_ai/chatgpt.py
Comment (rbs-jacob) [inline ofrak_ai/ofrak_ai/chatgpt.py]:
```suggestion
    temperature: float = 1.0
```

**Comment by rbs-jacob** [comment]
> Also, is there some standard way to more explicitly emphasize to ChatGPT where the prompt end and the data begins, so that there is less chance ChatGPT would react to instruction-like text in the actual code string?

@ANogin I suspect that if this were doable, there wouldn't be many prompt leaks, [including the original ChatGPT prompt](https://twitter.com/goodside/status/1598253337400717313) leaked from OpenAI itself and the [original Bing AI prompt](https://arstechnica.com/information-technology/2023/02/ai-powered-bing-chat-spills-its-secrets-via-prompt-injection-attack/). There might be a way to try and convince it to ignore future instructions, but I doubt there is anything airtight.

It's crazy that after decades of issues stemming from code and data being sent over the same input channel, we're solidifying at least some more years of that with LLMs that can't (currently, as far as I know) differentiate between the two.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
Some quick stuff, more to come...

File: ofrak_ai/ofrak_ai/sassy_string_modifier.py
Comment (EdwardLarson) [inline ofrak_ai/ofrak_ai/sassy_string_modifier.py]: All the fields of `SassyStringModifierConfig` have default values, so we can easily provide a default config for this modifier.

```suggestion
    async def modify(self, resource: Resource, config: SassyStringModifierConfig = SassyStringModifierConfig()):
```

File: ofrak_ai/ofrak_ai/sassy_string_modifier.py
Comment (EdwardLarson) [inline ofrak_ai/ofrak_ai/sassy_string_modifier.py]: This is another prompt constraint that could be validated.

File: ofrak_ai/Dockerfile
Comment (EdwardLarson) [inline ofrak_ai/Dockerfile]: Move these to a requirements.txt file

File: ofrak_ai/ofrak_ai/sassy_string_modifier.py
Comment (marczalik) [reply ofrak_ai/ofrak_ai/sassy_string_modifier.py]: @EdwardLarson which of the 2 constraints are you referring to?

By the way, I had issues when attempting to make ChatGPT follow the "no spaces" constraint. It seemed to not consider its commentary as part of its response, so if I sent it a request to sassify a string and respond without spaces, it would repeatedly send back something like
```
"Oh honey, ThisSassyStringHasNoSpaces#KillingIt"
```
which is why I ended up going the route of forcing the constraint by taking the max length word from the response as the result.

I don't have a good idea for how to validate the format specifier order. I suppose I could scan and make a list from the input string of anything that looks like a known specifier, then do the same for the output and compare. This could get messy when considering all the possible potential modifiers. Thoughts?

File: ofrak_ai/ofrak_ai/chatgpt.py
Comment (marczalik) [reply ofrak_ai/ofrak_ai/chatgpt.py]: I haven't played around with changing the model much, but it would certainly change the cost of the query and likely the speed as well. From what I've read, it could also affect things like how well ChatGPT follows constraints. For instance, GPT-4 is reportedly better at following constraints than 3.5. I think testing with different models would be worthwhile. One consequence of that would be: how do we want to present the results of that testing to potential users of this modifier? Presumably they will be making the choice of which model to use based on their own cost-benefit analysis.

File: ofrak_ai/Dockerfile
Comment (marczalik) [reply ofrak_ai/Dockerfile]: Removed as unnecessary

File: ofrak_ai/ofrak_ai/sassy_string_modifier.py
Comment (marczalik) [reply ofrak_ai/ofrak_ai/sassy_string_modifier.py]: @EdwardLarson Following up on the clarification to this feedback. Could you please specify which constraint should be validated in the response?

Comment (EdwardLarson) [reply ofrak_ai/ofrak_ai/sassy_string_modifier.py]: The one about C string format specifiers, it should be pretty doable to check for. And bad formatting could break the target binary if mishandled.

File: ofrak_ai/ofrak_ai/chatgpt.py
Comment (rbs-jacob) [reply ofrak_ai/ofrak_ai/chatgpt.py]: I think making this an enum instead of a string (with the same default value as now), and including a docstring that briefly mentions some trade-offs. Probably doesn't need to be too extensive.

Part of what I'm thinking about is that in the context of [this PR](https://github.com/redballoonsecurity/ofrak/pull/287), it would be easier for a user to fill out this config if they could select a model option from a drop-down instead of needing to know which magic string to put in.

Comment (marczalik) [reply ofrak_ai/ofrak_ai/chatgpt.py]: Makes total sense, I'll do that

File: ofrak_ai/ofrak_ai/sassy_string_modifier.py
Comment (EdwardLarson) [inline ofrak_ai/ofrak_ai/sassy_string_modifier.py]: This can just be compiled once (as a global, i.e. `SPECIFIER_PATTERN = re.compile(...)`)

File: ofrak_ai/ofrak_ai/chatgpt.py
Comment (EdwardLarson) [inline ofrak_ai/ofrak_ai/chatgpt.py]: This is unused and I don't see it being used in the future. Any analysis we use ChatGPT for should have a specific attributes type associated with it.

File: ofrak_ai/ofrak_ai/exponential_backoff.py
Comment (EdwardLarson) [inline ofrak_ai/ofrak_ai/exponential_backoff.py:50]:
```suggestion
    @functools.wraps
    async def wrapper(*args, **kwargs):
```
(You'll need to add the import as well)
This is generally a good thing to throw on any function decorator/wrapper: https://stackoverflow.com/questions/308999/what-does-functools-wraps-do

File: ofrak_ai/ofrak_ai/sassy_string_modifier.py
Comment (EdwardLarson) [inline ofrak_ai/ofrak_ai/sassy_string_modifier.py]: This is really a default value for the config and should be in the `SassyStringModifierConfig` declaration, but mutable defaults are a bit of a pain, so I see why it's in here.

File: ofrak_ai/ofrak_ai/chatgpt.py
Comment (marczalik) [reply ofrak_ai/ofrak_ai/chatgpt.py]: @EdwardLarson  this is used by the program and function analyzers that I haven't pushed yet. I can remove it from this PR. 

If I understand you correctly about the specific attribute types, I think that may be difficult to accomplish given the probabilistic nature of chatgpt's output. In many cases it will use a similar structure between responses, but (to varying degrees of frequency) it can and will break that structure. That's why at the moment I'm treating responses as one text blob rather than parsing them into pieces and assigning to specific attributes.

**Review by EdwardLarson — APPROVED** [review]

**Review by rbs-jacob — APPROVED** [review]
My proposed changes seem to be addressed, so I approve pending Edward's approval as well.

**Pull request merged**
